### PR TITLE
feat: automatic image extraction from docker-compose file

### PR DESCRIPTION
**Pull request by GoethalsRobbe**
Hi

As explained in [this post](https://hub.mender.io/t/docker-compose-images-in-github-ci-cd/6708), I need the app-gen to extract the images from the docker-compose file instead of passing them as arguments.
These changes would allow to do so.

@merlin-northern Please review and comment, thanks!

KR
Robbe

### CODE REVIEW

File: gen/app-gen [hunk 1]
```diff
@@ -208,9 +208,18 @@ if [ -z "${manifests_dir}" ]; then
     show_help_and_exit_error
 fi
```
Comment (merlin-northern) [inline gen/app-gen:210]: this modification is only valid for `"docker-compose"` orchestrator and so you have to check if `"$orchestrator" == "docker-compose"`

File: gen/app-gen [hunk 2]
```diff
@@ -208,9 +208,18 @@ if [ -z "${manifests_dir}" ]; then
...
 
-if [ ${#images[@]} -lt 1 ]; then
-    echo "No images specified. Aborting." >&2
-    show_help_and_exit_error
+if [ ${#images[@]} -lt 3 ]; then
```
Comment (merlin-northern) [inline gen/app-gen:211]: why `3`?

File: gen/app-gen [hunk 3]
```diff
@@ -208,9 +208,18 @@ if [ -z "${manifests_dir}" ]; then
...
+    echo "No specific images specified. Will try to extract from docker-compose.yaml file." >&2
+
+    while IFS= read -r img; do
+        images+=("$img")
+        images+=("$img")
+    done <<< $( docker compose --project-directory "$manifests_dir" config --images)
```
Comment (merlin-northern) [inline gen/app-gen:217]: thank you it is a lovely feature you are proposing.
you do not have to use _here strings_ and play with `IFS`, you can:

```suggestion
while read -r img; do
        images+=("$img")
        images+=("$img")
done < <( docker compose --project-directory "$manifests_dir" config --images)
```

and finally one day you could send the patch to accept "current" and "new" `manifests_dir` -- so you could modify your code to generate deltas based on the manifests!
I like your idea so much, that I would also think about doing the same for the `k8s` case.

File: gen/app-gen [hunk 4]
```diff
@@ -208,9 +208,18 @@ if [ -z "${manifests_dir}" ]; then
...
+
+    if [ ${#images[@]} -lt 3 ]; then
```
Comment (merlin-northern) [inline gen/app-gen:219]: why `3`?

**Review by merlin-northern — CHANGES_REQUESTED** [review]
thank you for the lovely feature. I am requesting changes to this patch because it will break the `k8s` orchestrator support.

File: gen/app-gen (hunk 2 above)
Comment (GoethalsRobbe) [reply gen/app-gen:211]: I first declared `declare -a images=("")` and `declare -a images_shas=("")` which caused the number of elements to be 1 already. I changed it to be empty lists but forget to adjust the minimum number of elements here. It's an obvious error from my side :wink:

File: gen/app-gen (hunk 3 above)
Comment (GoethalsRobbe) [reply gen/app-gen:217]: Thanks for the suggestion, works as intended. Will have to take a look on the mechanism behind `IFS` :+1:

**Review by merlin-northern — APPROVED** [review]
looks good. thank you. it is a nice feature.

**Pull request merged**